conan/tools/scm/git.py

```diff
@@ -0,0 +1,85 @@
+import os
+
+from conan.tools.files import chdir
+from conans.errors import ConanException
+from conans.util.files import mkdir
+from conans.util.runners import check_output_runner
+
+
+class Git(object):
+    def __init__(self, conanfile, folder):
+        self._conanfile = conanfile
+        self.folder = folder
+
+    def _run(self, cmd):
+        with chdir(self._conanfile, self.folder):
+            return check_output_runner("git {}".format(cmd)).strip()
+
+    def get_commit(self):
+        try:
+            # commit = self._run("rev-parse HEAD") For the whole repo
+            # This rev-list knows to capture the last commit for the folder
+            commit = self._run('rev-list HEAD -n 1 -- "{}"'.format(self.folder))
+            return commit
+        except Exception as e:
+            raise ConanException("Unable to get git commit in '%s': %s" % (self.folder, str(e)))
+
+    def get_remote_url(self, remote="origin"):
+        remotes = self._run("remote -v")
+        for r in remotes.splitlines():
+            name, url = r.split(maxsplit=1)
+            if name == remote:
+                url, _ = url.rsplit(None, 1)
+                if os.path.exists(url):  # Windows local directory
+                    url = url.replace("\\", "/")
+                return url
+
+    def commit_in_remote(self, commit, remote="origin", branch="master"):
+        if not remote or not branch:
+            return False
+        try:
+            branches = self._run("branch -r --contains {}".format(commit))
+            return "{}/{}".format(remote, branch) in branches
+        except Exception as e:
+            raise ConanException("Unable to check remote commit in '%s': %s" % (self.folder, str(e)))
+
+    def is_dirty(self):
+        status = self._run("status -s").strip()
+        return bool(status)
+
+    def get_url_commit(self, remote="origin", branch="master"):
```

Comment:
I think we should change this default to `main` also since git>2.28 there's the i`nit.defaultBranch` configuration option that we may check.

Reply:
I don't like the name `get_url_commit`, it reads like `get the URL of the commit` and makes no sense.
About the `remote` and `branch` defaults, maybe we should make it positional (not defaulted)? I would avoid extra commands like the `git config --get init.defaultBranch` suggested, even more, if they depend on the git version.

Reply:
Maybe we can avoid using the branch and just use the commit hash? Do we need the branch name?

Reply:
Agree, need to find a better name.

In any case, the branch needs to be worked out. It shouldn't be necessary, and indeed is very problematic for user flows like branching, PRs, in which the branch will not be "main" or "master". So the commit should be looked for directly in the remote without a branch hint, I'll investigate if this is possible.

Reply:
I have changed the name to ``get_url_and_commit()`` and removed the ``branch`` argument, now checking if the commit is in any branch. Added a test with an actual branch pushed, to see it still works.